galera/src/ist.cpp

```diff
@@ -395,6 +403,88 @@ galera::ist::Receiver::prepare(wsrep_seqno_t const first_seqno,
 }
 
 
+class SocketWatchdog
+{
+    public:
+        explicit SocketWatchdog(std::function<void()> onExpire, unsigned int timeoutMs = 10000)
+            : eventCbFn_(onExpire)
+            , active_(false)
+            , alive_(true)
+            , expire_cnt_(timeoutMs/10)
+            , mtx_()
+            , cv_()
+            , t_([this]() {
+
+              bool aliveSnapshot = alive_;
+
+              while(aliveSnapshot) {
+                bool activeSnapshot = false;
+                aliveSnapshot = false;
+                int counter = 0;
+
+                {
+                  std::unique_lock<std::mutex> lock(mtx_);
+                  while(!active_) cv_.wait(lock);
+                  activeSnapshot = active_;
+                  aliveSnapshot = alive_;
+                  counter = expire_cnt_;
+                }
+
+                while (activeSnapshot && aliveSnapshot) {
+                    if (counter == 0) {
+
+                        eventCbFn_();
+
+                        std::unique_lock<std::mutex> lock(mtx_);
+                        active_ = false;
+                        break;
+                    }
+
+                    {
+                        std::unique_lock<std::mutex> lock(mtx_);
+                        cv_.wait_for(lock, std::chrono::milliseconds(10));
+                        activeSnapshot = active_;
+                        aliveSnapshot = alive_;
+                    }
+                    --counter;
+                }
+              }
+          }) { }
+
+        ~SocketWatchdog() {
+            {
+                std::unique_lock<std::mutex> lock(mtx_);
+                alive_ = false;
+                active_ = true;
+                cv_.notify_one();
+            }
+            t_.join();
+
+        }
+
+        void restart() {
```

Comment:
I'm not sure if the name of this function is good: if the loop is currently running, it won't restart it, only if it is currently stopped.

While unlikely, it also seems possible that a quick stop(); restart(); sequence will happen without the thread noticing it, continuing the current counter instead of restarting it.